src/main/java/ecdar/backend/BackendDriver.java

```diff
@@ -166,18 +138,101 @@ public void onNext(Empty value) {
 
             @Override
             public void onError(Throwable t) {
+                newConnection.close();
+                startedEngineConnections.remove(newConnection);
             }
 
             @Override
             public void onCompleted() {
-                setConnectionAsAvailable(newConnection);
+                if (startedEngineConnections.contains(newConnection)) setConnectionAsAvailable(newConnection);
             }
         };
 
         newConnection.getStub().withDeadlineAfter(responseDeadline, TimeUnit.MILLISECONDS)
                 .updateComponents(componentsBuilder.build(), observer);
     }
 
+    /**
+     * Starts a process, creates an EngineConnection to it, and returns that connection
+     *
+     * @param engine to run in the process
+     * @return an EngineConnection to a local engine running in a Process or null if all ports are already in use
+     */
+    private EngineConnection startAndGetConnectionToEngineProcess(final Engine engine) {
```

Comment:
I think the name ``startEngine`` would be better. Maybe this function could also fit better in the Engine class and then the name ``start`` could be used.

Reply:
The naming was to make sure that the functionality was understood, but I think I went a bit far :laughing:

Starting the connections within the `Engine` might be a great approach. It is currently a pure data object and moving this logic into it would change the structure of the surrounding code a bit. 

> :information_source: Just a reminder, the `Engine` class is not a single running instance of an engine, but an abstraction of the information needed to start such instances.

Based on your suggestion and some consideration, I propose the following directions:

### 1) Move all handling of connections (creation, storage, and termination) into the `Engine` class
Each `Engine` would become an entry point to the associated connections. This would move 150-200 lines of business logic from the `BackendDriver` into the `Engine`. I am not sure if that is what we want, as the initial idea of the `BackendDriver` was to facilitate all the gRPC channel construction, communication, and termination. This way, the driver would only handle the "sending queries over the connections" side of things

### 2) Just move creation into the `Engine` class
100-130 lines of business code concerning the creation of `EngineConnection`s are moved from the `BackendDriver` to the `Engine`. When a new connection is wanted, the engine constructs one and forgets about it.